### PR TITLE
feat: implement textinput

### PR DESCRIPTION
**Pull request by TranQuangPhi**
https://user-images.githubusercontent.com/105619840/222658015-8faa77b9-5606-49ac-a2dc-4b112d1965a6.mov

https://user-images.githubusercontent.com/105619840/222657965-451807d1-512b-4e67-afa1-2f0181170fe6.mov

### CODE REVIEW

File: src/TextInput/TextInput.tsx [hunk 1]
```diff
@@ -0,0 +1,323 @@
...
+  height: ${props => (props.multiline ? metrics.massive : metrics.xxxl)};
+  min-height: 40;
+  margin-top: ${metrics.tiny};
+  margin-bottom: ${metrics.tiny};
+  border-bottom-width: 1;
+  border-bottom-color: ${colors.black};
+`
+
+const Title = styled.Text`
+  font-size: 16;
```
Comment (loc-nguyenthien) [inline src/TextInput/TextInput.tsx:299]: use `metrics`

File: src/TextInput/TextInput.tsx [hunk 2]
```diff
@@ -0,0 +1,323 @@
...
+`
+
+const InputContainer = styled.View<TextInputContainerProps>`
+  flex-direction: row;
+  align-items: center;
+  height: ${props => (props.multiline ? metrics.massive : metrics.xxxl)};
+  min-height: 40;
+  margin-top: ${metrics.tiny};
+  margin-bottom: ${metrics.tiny};
+  border-bottom-width: 1;
```
Comment (loc-nguyenthien) [inline src/TextInput/TextInput.tsx:294]: use `metrics`

File: src/TextInput/TextInput.tsx [hunk 3]
```diff
@@ -0,0 +1,323 @@
...
+  editable?: boolean
+
+  /** Styling for Input Component Container */
+  inputContainerStyle?: StyleProp<ViewStyle>
+
+  /** Style for Input Component */
+  inputStyle?: StyleProp<TextStyle>
+
+  /** Value of the text input */
+  value?: string
```
Comment (loc-nguyenthien) [inline src/TextInput/TextInput.tsx:35]: I think this should be a required prop

File: src/TextInput/TextInput.tsx [hunk 4]
```diff
@@ -0,0 +1,323 @@
...
+  icon?: ImageSourcePropType
+
+  /** Displays Icon size need to be used along with `icon` prop */
+  iconSize?: number
+
+  /** Displays Icon color need to be used along with `icon` prop */
+  iconColor?: string
+
+  /** Determines how to resize the image when the frame doesn't match the raw */
+  resizeMode?: ImageResizeMode
```
Comment (loc-nguyenthien) [inline src/TextInput/TextInput.tsx:59]: Change this to `iconResizeMode`

File: src/TextInput/TextInput.tsx [hunk 5]
```diff
@@ -0,0 +1,323 @@
...
+  returnKeyType?: ReturnKeyTypeOptions
+
+  /**
+   * If true, the text input obscures the text entered so that sensitive text like passwords stay secure.
+   * The default value is false.
+   */
+  secureTextEntry?: boolean
+
+  /** Callback that is called when the text input's text changes. Changed text is passed as an argument to the callback handler */
+  onChangeText?: (text: string) => void
```
Comment (loc-nguyenthien) [inline src/TextInput/TextInput.tsx:145]: I think this should be a required prop

**Review by loc-nguyenthien — CHANGES_REQUESTED** [review]

File: src/TextInput/TextInput.tsx (hunk 3 above)
Comment (loc-nguyenthien) [reply src/TextInput/TextInput.tsx:35]: not need

File: src/TextInput/TextInput.tsx (hunk 5 above)
Comment (loc-nguyenthien) [reply src/TextInput/TextInput.tsx:145]: not need

File: src/TextInput/TextInput.tsx [hunk 6]
```diff
@@ -0,0 +1,323 @@
...
+  iconPosition?: 'left' | 'right'
+
+  /**
+   * Can tell TextInput to automatically capitalize certain characters.
+   *    characters: all characters,
+   *    words: first letter of each word
+   *    sentences: first letter of each sentence
+   *    none: don't auto capitalize anything
+   */
+  autoCapitalize?: 'none' | 'sentences' | 'words' | 'characters'
```
Comment (anhle10051996) [inline src/TextInput/TextInput.tsx:83]: separate  type

File: src/TextInput/TextInput.tsx [hunk 7]
```diff
@@ -0,0 +1,323 @@
...
+
+  /** Vertically align text when `multiline` is set to true */
+  textAlignVertical?: 'auto' | 'top' | 'bottom' | 'center'
```
Comment (anhle10051996) [inline src/TextInput/TextInput.tsx:86]: separate type

File: src/TextInput/TextInput.tsx [hunk 8]
```diff
@@ -0,0 +1,323 @@
...
+        )}
+        <InputContainer multiline={multiline} style={inputContainerStyle}>
+          {iconPosition === 'left' && renderedIcon}
+          <TextInputComponent
+            ref={ref}
+            as={Input}
+            value={value}
+            style={inputStyle}
+            editable={editable}
+            underlineColorAndroid="transparent"
```
Comment (anhle10051996) [inline src/TextInput/TextInput.tsx:252]: so we not see the underline on android anymore ?

File: src/TextInput/TextInput.tsx [hunk 9]
```diff
@@ -0,0 +1,323 @@
...
+  size,
+  color,
+  resizeMode,
+  iconContainerStyle,
+  iconStyle,
+}) => (
+  <IconWrapper style={iconContainerStyle}>
+    <Icon source={source} size={size} color={color} resizeMode={resizeMode} style={iconStyle} />
+  </IconWrapper>
+)
```
Comment (loido) [inline src/TextInput/TextInput.tsx:182]: can you write an Icon component used for the project?
The Button component can use the Icon component

**Review by loido — CHANGES_REQUESTED** [review]

File: src/TextInput/TextInput.tsx [hunk 10]
```diff
@@ -0,0 +1,323 @@
...
+  ImageResizeMode,
+  ImageStyle,
+  KeyboardTypeOptions,
+} from 'react-native'
+import {metrics} from '../helpers/metrics'
+import {Icon} from '../Icon/Icon'
+import styled from 'styled-components/native'
+import {colors} from '../helpers/colors'
+
+export interface TextInputProps {
```
Comment (loido) [inline src/TextInput/TextInput.tsx:21]: can you extend TextInputProps from react-native?

import {TextInputProps} from 'react-native';

**Review by loido — CHANGES_REQUESTED** [review]

File: src/TextInput/TextInput.stories.tsx [hunk 11]
```diff
@@ -11,29 +10,70 @@ export default {
...
+      <View style={{marginTop: 50}} />
+      <TextInput.Flat
+        label={'Icon left'}
+        value={text}
+        onChangeText={setText}
+        autoCapitalize={'words'}
+        autoCorrect
+        leftComponent={<TextInput.Icon source={require('../assets/folder-search.png')} />}
+        containerStyle={{marginHorizontal: 20}}
+        errorText={'sadasdsadsa'}
```
Comment (loc-nguyenthien) [inline src/TextInput/TextInput.stories.tsx:26]: Use a more descriptive content for these

File: src/TextInput/TextInput.tsx [hunk 12]
```diff
@@ -268,56 +228,45 @@ export const TextInput = forwardRef<Input, TextInputProps>(
...
+
+const Container = styled.View({
+  flex: 1,
+})
+
+const InputContainer = styled.View<TextInputContainerProps>(props => ({
+  flexDirection: 'row',
+  alignItems: 'center',
+  height: props.multiline ? metrics.massive : metrics.huge,
+  borderWidth: 1,
```
Comment (loc-nguyenthien) [inline src/TextInput/TextInput.tsx:247]: Create an `border` in `metrics`

File: src/TextInput/TextInput.tsx [hunk 13]
```diff
@@ -268,56 +228,45 @@ export const TextInput = forwardRef<Input, TextInputProps>(
...
+  borderRadius: 6,
+  marginVertical: metrics.tiny,
+  borderColor: colors.black,
+  minHeight: metrics.huge,
+  paddingLeft: metrics.xxs,
+}))
+
+const Title = styled.Text({
+  fontSize: responsiveFont(16),
+  color: '#8B8B8B',
```
Comment (loc-nguyenthien) [inline src/TextInput/TextInput.tsx:257]: use `colors`

File: src/TextInput/TextInputFlat.tsx [hunk 14]
```diff
@@ -0,0 +1,203 @@
...
+                style={[
+                  styles.labelContainer,
+                  {
+                    left: leftComponent ? moveToLeft : metrics.xxs,
+                    top: moveToTop,
+                    fontSize: reduceFontSize,
+                    color: isFocused ? colors.primary : colors.placeHolderText,
+                  },
+                  labelStyle,
+                ]}
```
Comment (loc-nguyenthien) [inline src/TextInput/TextInputFlat.tsx:131]: should use `StyleSheet.flatten` for this, if you pass custom `labelStyle` in here

File: src/TextInput/TextInputFlat.tsx [hunk 15]
```diff
@@ -0,0 +1,203 @@
...
+})
+
+const ContentContainer = styled.View({
+  flex: 1,
+})
+
+const TextInputContainer = styled.View<TextInputContainerProps>(props => ({
+  flexDirection: 'row',
+  alignItems: 'center',
+  borderWidth: 1,
```
Comment (loc-nguyenthien) [inline src/TextInput/TextInputFlat.tsx:191]: same above, use `metrics` for `border`

**Review by loc-nguyenthien — CHANGES_REQUESTED** [review]

File: src/TextInput/TextInput.tsx (hunk 1 above)
Comment (TranQuangPhi) [reply src/TextInput/TextInput.tsx]: done

File: src/TextInput/TextInput.tsx (hunk 2 above)
Comment (TranQuangPhi) [reply src/TextInput/TextInput.tsx]: done

File: src/TextInput/TextInput.tsx (hunk 4 above)
Comment (TranQuangPhi) [reply src/TextInput/TextInput.tsx]: ImageResizeMode type is defined by React Nativ and I think we should not change its name

File: src/TextInput/TextInput.tsx (hunk 6 above)
Comment (TranQuangPhi) [reply src/TextInput/TextInput.tsx]: removed

File: src/TextInput/TextInput.tsx (hunk 7 above)
Comment (TranQuangPhi) [reply src/TextInput/TextInput.tsx]: removed

File: src/TextInput/TextInput.tsx (hunk 9 above)
Comment (TranQuangPhi) [reply src/TextInput/TextInput.tsx]: done

File: src/TextInput/TextInput.tsx (hunk 10 above)
Comment (TranQuangPhi) [reply src/TextInput/TextInput.tsx]: done

File: src/TextInput/TextInput.stories.tsx (hunk 11 above)
Comment (TranQuangPhi) [reply src/TextInput/TextInput.stories.tsx:26]: done

File: src/TextInput/TextInput.tsx (hunk 13 above)
Comment (TranQuangPhi) [reply src/TextInput/TextInput.tsx:181]: done

File: src/TextInput/TextInputFlat.tsx (hunk 15 above)
Comment (TranQuangPhi) [reply src/TextInput/TextInputFlat.tsx:193]: done

**Review by loc-nguyenthien — APPROVED** [review]

**Review by baonguyenhsts — COMMENTED** [review]
Does your component support animated title?
Example like this one



If not, could you consider adding that feature?

**Comment by TranQuangPhi** [comment]
> 

asdas

**Comment by TranQuangPhi** [comment]
> Does your component support animated title? Example like this one
> 
>  [ 68747470733a2f2f692e696d6775722e636f6d2f4e61314b4c49452e676966 ](https://camo.githubusercontent.com/759e9531e1cc872a445550e259a7d06d690bece413fcd5e8ed4968ce740480c3/68747470733a2f2f692e696d6775722e636f6d2f4e61314b4c49452e676966) [ ](https://camo.githubusercontent.com/759e9531e1cc872a445550e259a7d06d690bece413fcd5e8ed4968ce740480c3/68747470733a2f2f692e696d6775722e636f6d2f4e61314b4c49452e676966)
> 
> If not, could you consider adding that feature?

Let me confirm whether you are referring to the title animation jumping or the title animation moving to the top but remaining inside the border. Which one did you mention?

**Comment by baonguyenhsts** [comment]
> > Does your component support animated title? Example like this one
> > 
> >       
> >         [
> >           
> >         
> >             68747470733a2f2f692e696d6775722e636f6d2f4e61314b4c49452e676966
> >           ](https://camo.githubusercontent.com/759e9531e1cc872a445550e259a7d06d690bece413fcd5e8ed4968ce740480c3/68747470733a2f2f692e696d6775722e636f6d2f4e61314b4c49452e676966)
> >         
> >         
> >           
> >             
> >               
> >             
> >             
> >               
> >               
> >             
> >           
> >           [
> >             
> >               
> >             
> >           ](https://camo.githubusercontent.com/759e9531e1cc872a445550e259a7d06d690bece413fcd5e8ed4968ce740480c3/68747470733a2f2f692e696d6775722e636f6d2f4e61314b4c49452e676966)
> >         
> >       
> > If not, could you consider adding that feature?
> 
> Let me confirm whether you are referring to the title animation jumping or the title animation moving to the top but remaining inside the border. Which one did you mention?
> 
> the title animation jumping like below:

Yes, that's it

**Pull request closed without merging**